app/uk/gov/hmrc/apipublisher/connectors/APIScopeConnector.scala

```diff
@@ -47,12 +45,24 @@ class APIScopeConnector @Inject()(config: ApiScopeConfig, http: HttpClient)(impl
     http.POST[JsValue, Either[UpstreamErrorResponse, HttpResponse]](url, scopes)
       .map {
         case Right(_) => None
-        case Left(UpstreamErrorResponse(message, BAD_REQUEST, _, _)) => 
+        case Left(UpstreamErrorResponse(message, BAD_REQUEST, _, _)) =>
           Logger.debug(s"Failed request. POST url=$url: $message")
           Some(JsString(message))
         case Left(err) => throw err
       }
   }
+
+  def retrieveScopes(scopeKeys: Seq[String])(implicit hc: HeaderCarrier): Future[Option[JsValue]] = {
```

Comment:
Dislike passing JSON around the application.
It's fine as a transport mechanism between service but internally to the app it would be nice to _not_ pass JSON around.

In this case you only need the collection of scopes to be returned right?
And maybe an error should something goes wrong?

Reply:
We need collection of scope data array to be returned which includes key, name & description values of each scope and also it would be easy to compare two JsValue(s) directly to check if they are different with an == operator and hence returning it.

Reply:
I understand why this has been done and keeping the new code consistent with the old code is a good idea for the sake of others who come to this code later.

However, passing JsValues around is a bit nasty.

Especially since we already model the data you need in a case class and there is already a json formatter for it.

https://github.com/hmrc/api-publisher/blob/master/app/uk/gov/hmrc/apipublisher/models/PublisherRequest.scala#L107

It's not your fault in any way. The app was written oddly to begin with and you have to bolt on extra stuff in the least disruptive way possible. However, it is a good idea to always think about trying to tidy up this kind of stuff when you come across it.

As it is, the app has an internal structure that is a mix of Scala and JSON. I imagine passing JSON through the tiers was a convenience but it has coupled the internal representations of our data structures to the transport mechanism we use to talk to other services. Not great :(